File: 151003/Savitski/publisher/src/main/java/com/example/distributedcomputing/service/TagService.java
package com.example.distributedcomputing.service;

import com.example.distributedcomputing.exceptions.NotFoundException;
import com.example.distributedcomputing.mapper.TagMapper;
import com.example.distributedcomputing.model.entity.Tag;
import com.example.distributedcomputing.model.request.TagRequestTo;
import com.example.distributedcomputing.model.response.TagResponseTo;
import com.example.distributedcomputing.repository.TagRepository;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Min;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.validation.annotation.Validated;

@Service
@RequiredArgsConstructor
@Validated
public class TagService {
    private final TagRepository repository;
    private final TagMapper mapper;

    public Iterable<TagResponseTo> getAll() {
        return mapper.entityToDto(repository.findAll());
    }
    public TagResponseTo getById(@Min(0) Long id) {
        return repository.findById(id)
                .map(mapper::entityToDto)
                .orElseThrow();
    }
    public TagResponseTo save(@Valid TagRequestTo entity) {
        return mapper.entityToDto(repository.save(mapper.dtoToEntity(entity)));
    }
    public TagResponseTo update(@Valid TagRequestTo entity) {
        Long id = entity.id();
        Tag editor = mapper.dtoToEntity(entity);
        editor.setId(id);
        return mapper.entityToDto(repository.save(editor));
    }
    public void delete(@Min(0) Long id) {
        repository.findById(id).orElseThrow(() -> new NotFoundException(404L, "Not founded"));
        repository.deleteById(id);
    }
}

File: 151004/Bashlikov/rv_task04/publisher/src/main/kotlin/services/TagService.kt
package by.bashlikovvv.services

import by.bashlikovvv.api.dto.request.CreateTagDto
import by.bashlikovvv.api.dto.request.UpdateTagDto
import by.bashlikovvv.api.dto.response.TagDto

interface TagService {

    suspend fun create(createTagDto: CreateTagDto): TagDto?

    suspend fun getAll(): List<TagDto?>

    suspend fun getById(tagId: Long): TagDto?

    suspend fun update(tagId: Long, updateTagDto: UpdateTagDto): TagDto?

    suspend fun delete(tagId: Long): Boolean

}